### DISCUSSION TITLE
[Tutorial] A way to run CDP on TrueNAS Scale - K3s and Docker

### DISCUSSION THREAD

**Discussion by WiiGe** [stashapp/stash#4719] [General]
A way to use CDP scraper on TrueNAS Scale and install chromedp/headless-shell on TrueNAS Scale

People using stash-app on TrueNAS Scale like me may face the problem that no supported CDP app by truecharts or truenas , and now I found a way to properly run [chromedp/headless-shell
](https://github.com/chromedp/docker-headless-shell).

this can be done by custom-app:
![image](https://github.com/stashapp/stash/assets/8965675/a932906d-4952-4c70-b478-6414d535531b)

according to [docker hub page of chromedp/headless-shell](https://hub.docker.com/r/chromedp/headless-shell/), we start container with  `docker run -d -p 9222:9222 --rm --name headless-shell --shm-size 2G chromedp/headless-shell`, but this's not enough for TrueNAS Scale which run everything with K3s.

In order to run a docker image on K3s, we need to warp docker image with [custom-app](https://truecharts.org/charts/stable/custom-app/), fill the Container Repository with `chromedp/headless-shell` and Container Tag with `latest` so that truenas can pull the image for you.

![image](https://github.com/stashapp/stash/assets/8965675/e9380c95-7f4d-46bf-ae96-2823c9436121)

as for the docker args `docker run -d -p 9222:9222 --rm --name headless-shell --shm-size 2G chromedp/headless-shell`, we need to translate the args and env parameter into custom-app configuration. Under the Workload Settings section of custom-app configuration page, there will be a list of Extra Args, we put `--name headless-shell --shm-size 2G` here in custom-app way:

![image](https://github.com/stashapp/stash/assets/8965675/3025b83d-9bce-496e-af9b-0be555acb561)

but due to [this issue](https://github.com/Zenika/alpine-chrome/issues/109#issuecomment-1998061550), K3s will report an error of zygote: `ERROR:zygote_host_impl_linux.cc(262)] Failed to adjust OOM score of renderer with pid 28: Permission denied (13)`,  solution is simple - just offer extra args to disable it :`--headless --disable-gpu --no-sandbox  --no-zygote --disable-software-rasterizer --disable-dev-shm-usage`

`--remote-debugging-address=0.0.0.0 --remote-debugging-port=9222` is not needed because we will set port and ip on Networking and Services section for LoadBalancer (I am using 192.168.1.45 and 9222, change it for your own scenario):

the Networking section should be like this:
![image](https://github.com/stashapp/stash/assets/8965675/531f21a9-e4a6-4044-b8f4-2f707bb97554)

and, the final Extra Args page should be like this:
![image](https://github.com/stashapp/stash/assets/8965675/3824cd43-edb9-486c-af02-dbd37f9db85f)
and click Install button then your headless-shell will be good to go

the last step: just fill stash's CDP setting with `http://192.168.1.45:9222/json/version`(still, change ip:port for your own scenario), and now cdp should be running for you.  

An extra note is Ingress section of custom-app, you may want to use chromedp/headless-shell remotely (e.g. 
 accessing cdp by https://cdp.example.com/json/version), you need to modify the header of request with SetHeader function, see [Host header is specified and is not an IP address or localhost](https://github.com/chromedp/chromedp/issues/505). Obviously stash scraper do not support such operation, write your own scraper in this circumstances.

I just put the info I found together and share here, hope this page help more people who try to use CDP scraper on K3s-based OS.

**Comment by DogmaDragon** [comment]
Thank you. Would you be okay if I included this in documentation on https://docs.stashapp.cc?

**Reply by WiiGe** [reply]
sure！feel free to use, glad to see more stash user get helped

**Comment by Quixus** [comment]
This does not seem to work with 25.04 anymore. There are no fields to add the extra args. Would you add them in the custom command section and if so what is the command to add the extra args to? The Networking section also looks different in 25.04.

I'd highly appreciate if you updated the guide to the current version of Truenas Scale.

**Comment by WiiGe** [comment]
> I'd highly appreciate if you updated the guide to the current version of Truenas Scale.

Fine, it's actually much easier than TrueNAS Scale with K3s. As you can see, TrueNAS Scale removed K3s and turn to Docker now. So we can install [chromedp/headless-shell](https://hub.docker.com/r/chromedp/headless-shell/) in traditional Docker way.

we still use the [hromedp/headless-shell](https://hub.docker.com/r/chromedp/headless-shell/) recommended command: `docker run -d -p 9222:9222 --rm --name headless-shell --shm-size 2G chromedp/headless-shell` to start container in this guide update. The traditional Docker way to manage multiple containers is composing Docker compose file, for that we go [composerize](https://www.composerize.com/) to convert this commad into docker compose file(I name it with `headless_chromedp`, change with your favor):

```yaml
name: headless_chromedp
services:
  headless-shell:
    container_name: headless-shell
    image: chromedp/headless-shell
    ports:
      - '9222:9222'
    shm_size: 2G
```

note the section
```
ports:
      - '9222:9222'
```
it specified the access port 9222

save these content into a compose file - let's name it `chromedp_compose.yaml` and put it to TrueNAS directory: `/mnt/storage0/chromedp/chromedp_compose.yaml`

now we can turn to TrueNAS Scale and install chromedp App: click `Install via YAML` in `App - Discover Apps`
<img width="663" height="213" alt="image" src="https://github.com/user-attachments/assets/a8126a8f-d74a-47dd-a75b-146624e0545f" />
and `include` the docker compose file you just created in `/mnt/storage0/chromedp/chromedp_compose.yaml`, fill the App Name or you can't save your cahnge:

```yaml
include:
  - /mnt/storage0/chromedp/chromedp_compose.yaml
```

<img width="1169" height="582" alt="image" src="https://github.com/user-attachments/assets/77e89ec3-f0be-4d9c-89f6-242f194dd465" />

last step, click `Save` then job done:

<img width="1406" height="338" alt="image" src="https://github.com/user-attachments/assets/319ea45b-d5f3-4931-86e3-b455c82c5d4a" />

BTW, if you hate SSH and terminal, install a vscode-server or use vscode's Remote Explorer

**Comment by WiiGe** [comment]
and I'd highly appreciate if you sync this update to the original guide on https://docs.stashapp.cc/ @DogmaDragon

**Reply by DogmaDragon** [reply]
Hey, I will moving many of the community guide from the documenation to our community forum for easier access. 
I posted the updated version to the Community Wiki category https://discourse.stashapp.cc/t/run-cdp-on-truenas-scale/4041 and assigned the ownership to you. Since it's a wiki post, others can edit it too. 

Thank you for all the work on this.
